File: src/decoder.js
/**
 * AICL Decoder — AICL → original text
 *
 * Buffer-based scan with modifier support:
 *   - Base word symbol → flush buffer, start new word
 *   - Modifier symbol → apply transform to current buffer
 *   - Escape marker (\uE000) → flush buffer, next char is literal
 *   - Known symbol → flush buffer, expand pattern
 *   - Anything else → flush buffer, emit literal
 *
 * Must exactly invert encode(): decode(encode(x)) === x
 */

import { build, isModifier, getModifierTransform } from './dict.js';
import { codePoints, isPuaCodePoint, ESCAPE_MARKER, requireText } from './unicode.js';

/**
 * Decode AICL back into original text.
 * @param {string} aiclText
 * @param {object} [opts]
 * @param {boolean} [opts.steps=false] record expansion steps
 * @returns {{ output: string, steps?: object[], expansions?: number, literals?: number }}
 */
export function decode(aiclText, opts = {}) {
  requireText(aiclText, 'aiclText');
  const { symbolToPattern, modifierMap } = build();
  const steps = opts.steps ? [] : null;

  const chars = codePoints(aiclText);
  let output = '';
  let expansions = 0;
  let literals = 0;
  let i = 0;

  // Buffer holds the current word being reconstructed.
  // Modifiers append to this buffer. It's flushed when a non-modifier is encountered.
  let buffer = '';

  const flushBuffer = () => {
    if (buffer.length > 0) {
      output += buffer;
      buffer = '';
    }
  };

  while (i < chars.length) {
    const ch = chars[i];

    // MODIFIER: apply transform to buffer
    if (modifierMap.has(ch)) {
      const { transform } = modifierMap.get(ch);
      buffer = transform(buffer);
      if (steps) steps.push({ type: 'modifier', name: modifierMap.get(ch).name, pos: i });
      i++;
      continue;
    }

    // ESCAPE MARKER: flush buffer, next char is literal
    if (ch === ESCAPE_MARKER) {
      flushBuffer();
      const literal = chars[i + 1];
      if (literal === undefined) {
        // Dangling escape marker at end — emit it as-is (defensive).
        output += ch;
        literals++;
        i += 1;
      } else {
        output += literal;
        literals++;
        i += 2;
      }
      continue;
    }

    // KNOWN SYMBOL: flush previous buffer, start new buffer with pattern
    const pattern = symbolToPattern.get(ch);
    if (pattern !== undefined) {
      flushBuffer();
      buffer = pattern;  // hold in buffer so modifiers can transform it
      expansions++;
      if (steps) steps.push({ type: 'expand', symbol: ch, pattern, pos: i });
      i += 1;
      continue;
    }

    // LITERAL: flush buffer, emit char as-is
    flushBuffer();
    output += ch;
    literals++;
    if (steps) steps.push({ type: 'literal', char: ch, pos: i });
    i += 1;
  }

  // Flush any remaining buffer
  flushBuffer();

  const result = { output, expansions, literals };
  if (steps) result.steps = steps;
  return result;
}

/** Convenience: decode and return just the text. */
export function decodeToString(aiclText) {
  return decode(aiclText).output;
}

File: aicl/decoder.py
"""Stage 1 dictionary decoder."""

from __future__ import annotations

from typing import Any, Mapping

from .dict import build
from .unicode import ESCAPE_MARKER, code_points, require_text


def decode(aicl_text: str, opts: Mapping[str, Any] | None = None) -> dict[str, Any]:
    """Decode AICL text and optionally record expansion steps."""

    require_text(aicl_text, "aiclText")
    options = opts or {}
    steps_enabled = bool(options.get("steps", options.get("stepsEnabled", False)))
    steps: list[dict[str, Any]] | None = [] if steps_enabled else None

    data = build()
    symbol_to_pattern: Mapping[str, str] = data["symbolToPattern"]  # type: ignore[assignment]
    modifier_map: Mapping[str, Mapping[str, object]] = data["modifierMap"]  # type: ignore[assignment]
    chars = code_points(aicl_text)
    output: list[str] = []
    buffer = ""
    expansions = 0
    literals = 0
    index = 0

    def flush_buffer() -> None:
        nonlocal buffer
        if buffer:
            output.append(buffer)
            buffer = ""

    while index < len(chars):
        ch = chars[index]
        modifier = modifier_map.get(ch)
        if modifier is not None:
            transform = modifier["transform"]
            assert callable(transform)
            buffer = transform(buffer)
            if steps is not None:
                steps.append({"type": "modifier", "name": modifier["name"], "pos": index})
            index += 1
            continue

        if ch == ESCAPE_MARKER:
            flush_buffer()
            literal = chars[index + 1] if index + 1 < len(chars) else None
            if literal is None:
                output.append(ch)
                literals += 1
                index += 1
            else:
                output.append(literal)
                literals += 1
                index += 2
            continue

        pattern = symbol_to_pattern.get(ch)
        if pattern is not None:
            flush_buffer()
            buffer = pattern
            expansions += 1
            if steps is not None:
                steps.append({"type": "expand", "symbol": ch, "pattern": pattern, "pos": index})
            index += 1
            continue

        flush_buffer()
        output.append(ch)
        literals += 1
        if steps is not None:
            steps.append({"type": "literal", "char": ch, "pos": index})
        index += 1

    flush_buffer()
    result: dict[str, Any] = {"output": "".join(output), "expansions": expansions, "literals": literals}
    if steps is not None:
        result["steps"] = steps
    return result


def decode_to_string(aicl_text: str) -> str:
    return decode(aicl_text)["output"]


decodeToString = decode_to_string

__all__ = ["decode", "decodeToString", "decode_to_string"]
